text/2018-10-25-tikv-client-rust.md

```diff
@@ -0,0 +1,389 @@
+
+We choose to name the crate `tikv_client` to conform to the constraints presented to us by the Rust compiler. Cargo permits `tikv-client` and `tikv_client`, but `rustc` does not permit `tikv-client` so we choose to use `tikv_client` to reduce mental overhead.
+
+Choosing to seperate `tikv` and `client` helps potentially unfamiliar users to immediately understand the intent of the package. `tikvclient`, while understandable, is not immediately parsable by a human.
+
+All structures and functions will otherwise follow the [Rust API Guidelines](https://rust-lang-nursery.github.io/api-guidelines/), some of which will be enforced by `clippy`.
+
+## Installation
+
+To utilize the client programmatically, users will be able to add the `tikv-client` crate to their `Cargo.toml`'s dependencies. Then they must use the crate with `use tikv_client;`. Unfortunately due to Rust’s naming conventions this inconsistency is in place.
+
+To utilize the command line client, users will be able to install the binary via `cargo install tikv-client`. They will then be able to access the client through the binary `tikv-client`. If they wish for a different name they can alias it in their shell.
+
+## Usage
+
+## Two types of APIs
+TiKV provides two types of APIs for developers:
+- The Raw Key-Value API
+
+    If your application scenario does not need distributed transactions or MVCC (Multi-Version Concurrency Control) and only needs to guarantee the atomicity towards one key, you can use the Raw Key-Value API.
+
+- The Transactional Key-Value API
+
+    If your application scenario requires distributed ACID transactions and the atomicity of multiple keys within a transaction, you can use the Transactional Key-Value API.
+
+Generally, the Raw Key-Value API has higher throughput and lower latency compared to the Transactional Key-Value API. If distributed ACID transactions are not required, Raw Key-Value API is preferred over Transactional Key-Value API for better performance and ease of use.
+
+The client provides two types of APIs in two separate modules for developers to choose from.
+
+### The common data types
+
+- Key: raw binary data
+- Value: raw binary data
+- KvPair: Key-value pair type
+- KeyRange: Half-open interval of keys
+- Config: Configuration for client
+
+### Raw Key-Value API Basic Usage
+
+To use the Raw Key-Value API, take the following steps:
+
+1. Create an instance of Config to specify endpoints of PD (Placement Driver) and optional security config
+
+    ```rust
+    let config = Config::new(vec!["127.0.0.1:2379"]);
+    ```
+
+2. Create a Raw Key-Value client.
+
+    ```rust
+    let client = RawClient::new(&config);
+    ```
+
+3. Call the Raw Key-Value client methods to access the data on TiKV. The Raw Key-Value API contains following methods
+
+    ```rust
+    pub trait Client {
+        type AClient: Client;
+
+        fn get(&self, key: impl AsRef<Key>) -> Get<Self::AClient>;
```

Comment:
What is `Get<T>`?

Reply:
@Hoverbear Is it too verbose to include everything here? I mean the Definition of Get/BatchGet/Put/BatchPut/Transaction/Scanner etc.

Reply:
I don't have a strong opinion either way.

Reply:
Since they are builder of real requests which tends to be verbose and less interesting. I would prefer to keep them in real API document.